crates/pipeline-manager/src/compiler/main.rs

```diff
@@ -0,0 +1,268 @@
+use crate::api::error::ApiError;
+use crate::compiler::error::CompilerError;
+use crate::compiler::rust_compiler::{
+    perform_rust_compilation, rust_compiler_task, RustCompilationError,
+};
+use crate::compiler::sql_compiler::{
+    perform_sql_compilation, sql_compiler_task, SqlCompilationError,
+};
+use crate::compiler::util::recreate_dir;
+use crate::config::{CommonConfig, CompilerConfig};
+use crate::db::storage_postgres::StoragePostgres;
+use crate::db::types::common::Version;
+use crate::db::types::pipeline::PipelineId;
+use crate::db::types::program::ProgramConfig;
+use crate::db::types::tenant::TenantId;
+use crate::error::ManagerError;
+use crate::probe::Probe;
+use actix_files::NamedFile;
+use actix_web::{get, web, HttpRequest, HttpServer, Responder};
+use futures_util::join;
+use log::{error, info};
+use std::str::FromStr;
+use std::sync::Arc;
+use tokio::{spawn, sync::Mutex};
+use uuid::Uuid;
+
+/// Decodes the URL encoded parameter value as a string.
+fn decode_url_encoded_parameter(
+    param: &'static str,
+    value: Option<&str>,
+) -> Result<String, ManagerError> {
+    match value {
+        None => Err(ManagerError::from(ApiError::MissingUrlEncodedParam {
+            param,
+        })),
+        Some(checksum) => Ok(checksum.to_string()),
+    }
+}
+
+/// Validates the provided string is a hex-encoded SHA256 checksum.
+fn validate_is_sha256_checksum(s: &str) -> Result<(), ManagerError> {
+    if s.len() != 64 {
+        return Err(ApiError::InvalidChecksumParam {
+            value: s.to_string(),
+            error: format!("{} characters long instead of expected 64", s.len()),
+        }
+        .into());
+    };
+    for c in s.chars() {
+        if !c.is_ascii_digit()
+            && c != 'a'
+            && c != 'b'
+            && c != 'c'
+            && c != 'd'
+            && c != 'e'
+            && c != 'f'
+        {
+            return Err(ApiError::InvalidChecksumParam {
+                value: s.to_string(),
+                error: format!("character '{c}' is not hexadecimal (0-9, a-f)"),
+            }
+            .into());
+        }
+    }
+    Ok(())
+}
+
+/// Retrieves the binary executable.
+#[get("/binary/{pipeline_id}/{program_version}/{source_checksum}/{integrity_checksum}")]
+async fn get_binary(
+    config: web::Data<CompilerConfig>,
+    req: HttpRequest,
+) -> Result<impl Responder, ManagerError> {
+    // Retrieve URL encoded parameters
+    let path_parameters = req.match_info();
+    let pipeline_id =
+        decode_url_encoded_parameter("pipeline_id", path_parameters.get("pipeline_id"))?;
+    let program_version =
+        decode_url_encoded_parameter("program_version", path_parameters.get("program_version"))?;
+    let source_checksum =
+        decode_url_encoded_parameter("source_checksum", path_parameters.get("source_checksum"))?;
+    let integrity_checksum = decode_url_encoded_parameter(
+        "integrity_checksum",
```

Comment:
what's the difference between source and integrity checksum? can we document it somewhere?

Reply:
It's documented in the `PipelineDescr` (`pipeline-manager/src/db/types/pipeline.rs`):

```
    /// Combined checksum of all the inputs that influenced Rust compilation to a binary.
    pub program_binary_source_checksum: Option<String>,

    /// Checksum of the binary file itself.
    pub program_binary_integrity_checksum: Option<String>,
```

It's also written in the SQL migration.